core-bundle/src/Resources/contao/library/Contao/Controller.php

```diff
@@ -1513,6 +1519,218 @@ protected function eliminateNestedPages($arrPages, $strTable=null, $blnSorting=f
 		return $arrPages;
 	}
 
+	/**
+	 * Add an image to a template
+	 */
+	public static function addImageToTemplate(object $template, array $rowData, ?int $maxWidth = null, ?string $lightBoxGroupIdentifier = null, FilesModel $filesModel = null): void
+	{
+		// Helper: Create MetaData from the specified row data
+		$createMetaDataOverwriteFromRowData = static function (bool $interpretAsContentModel) use ($rowData)
+		{
+			if ($interpretAsContentModel)
+			{
+				/** @var ContentModel $contentModel */
+				$contentModel = (new \ReflectionClass(ContentModel::class))
+					->newInstanceWithoutConstructor();
+
+				// This will be null if `overwriteMeta` isn't set
+				return $contentModel->setRow($rowData)->getOverwriteMetaData();
+			}
+
+			// Manually create meta data that always contains certain properties (BC)
+			return new MetaData(
+				array(
+					MetaData::VALUE_ALT => $rowData['alt'] ?? '',
+					MetaData::VALUE_TITLE => $rowData['imageTitle'] ?? '',
+					MetaData::VALUE_URL => self::replaceInsertTags($rowData['imageUrl'] ?? ''),
+					'linkTitle' => $rowData['linkTitle'] ?: '',
+				)
+			);
+		};
+
+		// Helper: Create fallback template data with (mostly) empty fields (used if resource acquisition fails)
+		$createFallBackTemplateData = static function () use ($filesModel, $rowData)
+		{
+			$templateData = array(
+				'width' => null,
+				'height' => null,
+				'picture' => array(
+					'img' => array(
+						'src' => '',
+						'srcset' => '',
+					),
+					'sources' => array(),
+					'alt' => '',
+					'title' => '',
+				),
+				'singleSRC' => $rowData['singleSRC'],
+				'src' => '',
+				'linkTitle' => '',
+				'margin' => '',
+				'addImage' => true,
+				'addBefore' => true,
+				'fullsize' => false,
+			);
+
+			if (null !== $filesModel)
+			{
+				// Set empty meta data (this is actually a minor BC break)
```

Comment:
Why? And why is it ok to break BC here?

Reply:
This is only the fallback if the image resource cannot be found. Imo the old implementation should not have added meta data neither. It believe it just turned out this way because of how the meta data handling was written. We would need to duplicate part of the meta data logic (including the fallback language handling) in this method again just to support this edge case. I think that's not worth it.

I removed the additional note about it being a BC break in 30fa39480ce9ff7317e7bf28c0cdea96fbee93a8